### ISSUE TITLE
Test files do not have the correct filetype under (abnormal?) conditions

### ISSUE THREAD

**Issue by harleypig** [vim-perl/vim-perl#191]
test files are not being have the correct filetype detected If you have a .t file in a directory other than a directory called t and you have something like the following:

    BEGIN { use Test::Most; skip_all => 'author tests' unless $ENV{AUTHOR_TESTING} }
    eval { require Test::CheckDeps };
    plan skip_all => 'Test::CheckDeps required' if $@;
    ...

or even spread it out

    BEGIN {
      use Test::Most;
      skip_all =>  'author tests' unless $ENV{AUTHOR_TESTING}
    }
    eval { require Test::CheckDeps };
    plan skip_all => 'Test::CheckDeps required' if $@;
    ...

if you move the 'use' line so it's in the first column then the above filetype would be detected correctly, or if you have a use later in the file on the first column, the file it would be detected correctly as well.

$VIM/vim74/filetype.vim:2061-85 explains why this is happening.

Basically, the following checks are made.

    If the directory name the file is in is named 't', then it's a perl file, set and return.
    If the first line and the first column of the file contains a '#' and the first line contains the work perl, then it's a perl file, set and return.
    if there is a line that looks like '^use\s\s*\k' and can be found in the first 30ms of searching, it's a perl file, set and return.

It looks like this will not work for

    use
      Some::Module;

either.

We could mimic the checking done here and add somethin like

    if expand("%:e") == 't' && expand("%:p:h") =~ '/xt/'
      setf perl
      return 1
    endif

But I'm not sure where. would it a good idea to even include this?

**Comment by harleypig** [comment]
I was able to make the following work in after/ftdetect/perl.vim:

    function! IsPerlTestFile()
      if expand("%:e") == 't' && expand("%:p:h") =~ '/xt/'
        set filetype=perl
      endif
    endfunction
    
    au BufNewFile,BufRead *.t call IsPerlTestFile()

**Comment by hoelzro** [comment]
@harleypig I've seen that too, but I don't remember the filetype that was mistakenly used.  Do you know what it is? That might help to create a more encompassing solution.

**Comment by hoelzro** [comment]
After digging into the `filetype.vim` distributed with Vim 7.4.738, it looks like `.t` files are set to `nroff` or `tads` unless some specific conditions are met (one of which is residing within a directory named `t/`).  We should probably include `xt` and send a patch upstream.

**Comment by harleypig** [comment]
Can we reasonably assume that 'xt' anywhere in the path constitutes a test file? As opposed to the last directory in the path being 't' ...

I'm seeing a lot of cpan modules with xt/{author,release,smoke}/*.t files, and a few other arbitrary names in places. I've even seen deeper structures.

It's not a problem until and unless a test file is created with those particular parameters I mentioned, which is why I'm not sure if it's a valid thing to suggest  upstream (what else does this affect?) or to keep it in vim-perl, even if just as a contrib file.

**Comment by hoelzro** [comment]
@harleypig That seems reasonable.  Let's start with a contrib file, and merge it into `ftdetect` if enough people use it.
